File: Core/ModeRegistry.swift
import os

/// Associates bundle identifiers with `AppMode` conformers and resolves the
/// active mode for any given frontmost application.
///
/// Registration order does not matter — bundle IDs must be unique across all
/// registered modes. If a bundle ID is registered twice, a warning is logged
/// and the later registration wins.
///
/// Modes are registered in `AppDelegate` at startup:
/// ```swift
/// let registry = ModeRegistry(fallback: GlobalMode())
/// registry.register(SpotifyMode(), for: [BundleID.spotify])
/// registry.register(BrowserMode(), for: BundleID.chromiumBrowsers)
/// ```
final class ModeRegistry {

    // MARK: - Properties

    /// The mode used when no registered bundle ID matches the frontmost app.
    let fallback: any AppMode

    /// Flat lookup table built incrementally as modes are registered.
    private var bundleIDToMode: [String: any AppMode] = [:]

    // MARK: - Initialisation

    init(fallback: any AppMode) {
        self.fallback = fallback
    }

    // MARK: - Registration

    /// Registers an `AppMode` for one or more bundle identifiers.
    ///
    /// - Parameters:
    ///   - mode: The mode to activate when any of the given apps comes to front.
    ///   - bundleIdentifiers: One or more `CFBundleIdentifier` strings.
    func register(_ mode: any AppMode, for bundleIdentifiers: [String]) {
        for bundleID in bundleIdentifiers {
            if bundleIDToMode[bundleID] != nil {
                Logger.detection.warning(
                    "Bundle ID '\(bundleID)' is already registered — overwriting with '\(mode.id)'"
                )
            }
            bundleIDToMode[bundleID] = mode
        }
    }

    // MARK: - Resolution

    /// Returns the mode for the given bundle identifier, or `fallback` if none matches.
    func mode(for bundleIdentifier: String?) -> any AppMode {
        guard let bundleID = bundleIdentifier else { return fallback }
        return bundleIDToMode[bundleID] ?? fallback
    }

    /// Returns `true` if the given bundle identifier has a registered mode
    /// (excluding the fallback).
    func hasRegisteredMode(for bundleIdentifier: String?) -> Bool {
        guard let bundleID = bundleIdentifier else { return false }
        return bundleIDToMode[bundleID] != nil
    }
}

File: pocket_study_remote/core/mode_registry.py
"""ModeRegistry — maps bundle IDs to AppMode instances."""

from __future__ import annotations

import logging

from .app_mode import AppMode

logger = logging.getLogger(__name__)


class ModeRegistry:
    """
    Associates bundle identifiers with ``AppMode`` instances and resolves
    the active mode for any given frontmost application.

    Usage (in ``main.py``)::

        registry = ModeRegistry(fallback=GlobalMode())
        registry.register(SpotifyMode(), bundle_ids=[BundleID.SPOTIFY])
        registry.register(BrowserMode(), bundle_ids=BundleID.CHROMIUM_BROWSERS)
        registry.register(ObsidianMode(), bundle_ids=[BundleID.OBSIDIAN])

    Adding a new mode requires only a new ``register()`` call here — nothing
    else in the codebase changes.
    """

    def __init__(self, fallback: AppMode) -> None:
        """
        Args:
            fallback: The mode to use when no bundle ID matches.
                      Typically ``GlobalMode()``.
        """
        self._fallback = fallback
        self._bundle_to_mode: dict[str, AppMode] = {}

    # ------------------------------------------------------------------
    # Registration
    # ------------------------------------------------------------------

    def register(self, mode: AppMode, bundle_ids: list[str]) -> None:
        """
        Register an ``AppMode`` for one or more bundle identifiers.

        Args:
            mode:       The mode to activate when any of the given apps
                        comes to the front.
            bundle_ids: One or more ``CFBundleIdentifier`` strings.
        """
        for bundle_id in bundle_ids:
            if bundle_id in self._bundle_to_mode:
                logger.warning(
                    "Bundle ID %r is already registered — overwriting with %r",
                    bundle_id,
                    mode.id,
                )
            self._bundle_to_mode[bundle_id] = mode

    # ------------------------------------------------------------------
    # Resolution
    # ------------------------------------------------------------------

    def mode_for(self, bundle_identifier: str | None) -> AppMode:
        """
        Return the mode for the given bundle identifier, or the fallback
        if none is registered.
        """
        if bundle_identifier is None:
            return self._fallback
        return self._bundle_to_mode.get(bundle_identifier, self._fallback)

    @property
    def fallback(self) -> AppMode:
        return self._fallback
